proxy/src/main/java/com/wavefront/agent/handlers/DeltaCounterHandlerImpl.java

```diff
@@ -0,0 +1,185 @@
+            this::reportCache, reportIntervalSeconds, reportIntervalSeconds, TimeUnit.SECONDS);
+
+        String metricPrefix = entityType.toString() + "." + handle;
+        this.receivedCounter = registry.newCounter(new MetricName(metricPrefix, "", "received"));
+        this.reportedCounter = registry.newCounter(new MetricName(metricPrefix, "", "sent"));
+        this.rejectedCounter = registry.newCounter(new MetricName(metricPrefix, "", "rejected"));
+        estimatedSize = this.metricCache.estimatedSize();
+        Metrics.newGauge(new MetricName("cache", "", "estimated-size"),
+                new Gauge<Long>() {
+                    @Override
+                    public Long value() {
+                        return estimatedSize;
+                    }
+                });
+    }
+
+    private void reportCache() {
+        this.metricCache.asMap().forEach(this::reportDeltaCounter);
+    }
+
+    private void reportDeltaCounter(HostMetricTagsPair hostMetricTagsPair, AtomicDouble value) {
+        this.reportedCounter.inc();
+        if (value == null || hostMetricTagsPair == null) {return;}
+        double reportedValue = value.getAndSet(0);
+        if(reportedValue == 0) return;
+        String strPoint = metricToLineData(hostMetricTagsPair.metric, reportedValue,
+            System.currentTimeMillis(), hostMetricTagsPair.getHost(),
+            hostMetricTagsPair.getTags(), "wavefront-proxy");
+        getTask().add(strPoint);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    void reportInternal(ReportPoint point) {
+        if (validationConfig.get() == null) {
+            validatePoint(point, handle, Validation.Level.NUMERIC_ONLY);
+        } else {
+            validatePoint(point, validationConfig.get());
+        }
+        if (DeltaCounter.isDelta(point.getMetric())) {
+            String strPoint = serializer.apply(point);
+
+            refreshValidPointsLoggerState();
+
+            if ((logData || logPointsFlag) &&
+                (logSampleRate >= 1.0d || (logSampleRate > 0.0d && RANDOM.nextDouble() < logSampleRate))) {
+                // we log valid points only if system property wavefront.proxy.logpoints is true
+                // or RawValidPoints log level is set to "ALL". this is done to prevent introducing
+                // overhead and accidentally logging points to the main log.
+                // Additionally, honor sample rate limit, if set.
+                validPointsLogger.info(strPoint);
+            }
+
+            getReceivedCounter().inc();
+
+            double pvalue = Double.parseDouble(String.valueOf(point.getValue()));
+            receivedPointLag.update(Clock.now() - point.getTimestamp());
+            HostMetricTagsPair hostMetricTagsPair = new HostMetricTagsPair(point.getHost(),
+                point.getMetric(), point.getAnnotations());
+            metricCache.get(hostMetricTagsPair,
+                key -> new AtomicDouble(0)).getAndAdd(pvalue);
```

Comment:
This won't actually populate the cache, use metricCache.asMap().computeIfAbsent() instead.
Or, better yet, switch from AtomicDouble to Double and use compute():

```suggestion
            metricCache.asMap().compute(hostMetricTagsPair, (k, v) -> v == null ? pvalue : pvalue + v);
```

Reply:
nvm, metricCache.get() works too